### PR TITLE
[#32] workaround for https://github.com/openslide/openslide/issues/268

### PR DESCRIPTION
**Pull request by pavertomato**
tested on:

    ProductName: Mac OS X
    ProductVersion: 10.15
    BuildVersion: 19A602

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@ int main(int argc, char *argv[]) {
   int batch;
   int threads;
   bool debug;
+  bool openslideShift;
```
Comment (dzelemba) [inline src/main.cpp]: Maybe call this dropFirstRowAndColumn to make it more clear?

File: src/wsiToDcm.cpp [hunk 2]
```diff
@@ -277,12 +283,26 @@ int WsiToDcm::wsi2dcm(std::string inputFile, std::string outputFileMask,
...
         retileLevels,
         std::vector<double>(downsamples, downsamples + retileLevels + 1), tiled,
-        batchLimit, threads);
+        batchLimit, threads, openslideShift);
   } catch (int exception) {
     return 1;
   }
 }
 
+int WsiToDcm::wsi2dcm(std::string inputFile, std::string outputFileMask,
```
Comment (dzelemba) [inline src/wsiToDcm.cpp]: Can we take this opportunity to clean up this super long parameter list? Instead of having various overloads that are hard to follow, can we use some kind of request struct instead?

**Comment by pavertomato** [comment]
> Can we add a unit or integration test to cover this functionality?

i've added image comparing for image with offset

> Also mind explaining what kind of manual testing you did in the PR description?

i've checked what generated results not shifted in qupath

File: src/wsiToDcm.cpp (hunk 2 above)
Comment (pavertomato) [reply src/wsiToDcm.cpp]: done

File: src/main.cpp (hunk 1 above)
Comment (pavertomato) [reply src/main.cpp]: done

File: src/wsiRequest.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// limitations under the License.
+
+#ifndef SRC_WSIREQUEST_H_
+#define SRC_WSIREQUEST_H_
+
+#include <string>
+#include "src/enums.h"
+
+// Structure for wsi2dcm settings
+// inputFile - wsi input
```
Comment (dzelemba) [inline src/wsiRequest.h]: Put these comments on the actual member variables within the struct

File: src/wsiRequest.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+//          levels will be readed from wsi file
+// imageName, studyId, seriesId - required DICOM metadata
+// jsonFile - json file with additional DICOM metadata
+// downsample - for each level, downsample is size
+//              factor for each level
+// tiled - frame organization type
+//           true:TILED_FULL
+//           false:TILED_SPARSE
+// batchLimit - maximum frames in one file
+struct WsiRequest {
```
Comment (dzelemba) [inline src/wsiRequest.h]: It's fine to just define this struct in wsi2dcm.h. That way someone looking at functions in wsi2dcm doesn't have to jump to another file to understand the WsiRequest struct.

File: src/cWrapper.cpp [hunk 5]
```diff
@@ -34,8 +34,24 @@ int wsi2dcmJson(const char *inputFile, const char *outputFileMask,
                 const char *seriesId, const char *jsonFile, int retileLevels,
                 double *downsamples, bool tiled, int batchLimit, int threads,
                 bool debug) {
-  return wsiToDicomConverter::WsiToDcm::wsi2dcm(
-      inputFile, outputFileMask, frameSizeX, frameSizeY, compression, quality,
-      startOnLevel, stopOnLevel, imageName, studyId, seriesId, jsonFile,
-      retileLevels, downsamples, tiled, batchLimit, threads, debug);
+  WsiRequest request;
+  request.inputFile = inputFile;
```
Comment (dzelemba) [inline src/cWrapper.cpp]: What about WsiRequest request = {
  .inputFile = inputFile,
  .outputFileMask = outputFileMask,
  ...
}
?

Comment (pavertomato) [reply src/cWrapper.cpp]: well, this is C99 feature, id prefer avoid using of it

File: src/wsiRequest.h (hunk 4 above)
Comment (pavertomato) [reply src/wsiRequest.h]: done, but not sure how it would help since some who'll use this will have to jump to wsiToDcm.h, it's same as jump to wsiRequest.h

File: src/wsiRequest.h (hunk 3 above)
Comment (pavertomato) [reply src/wsiRequest.h]: done

File: src/cWrapper.cpp (hunk 5 above)
Comment (dzelemba) [reply src/cWrapper.cpp]: Ah yes good point, sounds good.

File: src/wsiToDcm.h [hunk 6]
```diff
@@ -22,51 +22,40 @@
...
+struct WsiRequest {
+  std::string inputFile = "";  // wsi input
+  std::string outputFileMask = "./";  // path to save generated files
+  int64_t frameSizeX = 500;
+  int64_t frameSizeY = 500;
+  DCM_Compression compression = JPEG;  // compression - jpeg, jpeg2000, raw
+  int32_t quality = 80;
+  int32_t startOnLevel = 0;   //
+  int32_t stopOnLevel = -1;   // limitation which levels to generate
+  std::string imageName = "image";  //
```
Comment (dzelemba) [inline src/wsiToDcm.h]: Describe this parameter... is this set as part of the DICOM metadata? If so, which tag?

File: src/wsiToDcm.h [hunk 7]
```diff
@@ -22,51 +22,40 @@
...
+  std::string studyId;  //
+  std::string seriesId;  // required DICOM metadata
+  std::string jsonFile = "";  // json file with additional DICOM metadata
+  int32_t retileLevels = 0;  // number of levels, levels == 0  means number of
+                            // levels will be readed from wsi file
+  double* downsamples;  // for each level, downsample is size
```
Comment (dzelemba) [inline src/wsiToDcm.h]: Given an example of how downsample works to make it a bit more clear.

File: src/wsiToDcm.h [hunk 8]
```diff
@@ -22,51 +22,40 @@
...
+                       // factor for each level
+  bool tiled;  // frame organization type
```
Comment (dzelemba) [inline src/wsiToDcm.h]: Can you provide a link that describes the differences between these two types?

File: src/wsiToDcm.h [hunk 9]
```diff
@@ -22,51 +22,40 @@
...
+               // true:TILED_FULL
+               // false:TILED_SPARSE
+  int32_t batchLimit;  //  maximum frames in one file
```
Comment (dzelemba) [inline src/wsiToDcm.h]: Describe what happens when limit is exceeded.

File: src/wsiToDcm.h [hunk 10]
```diff
@@ -22,51 +22,40 @@
...
+  int8_t threads = -1;
+  bool dropFirstRowAndColumn = false;
```
Comment (dzelemba) [inline src/wsiToDcm.h]: Add comment explaining why someone might use this.

File: src/wsiToDcm.h [hunk 11]
```diff
@@ -22,51 +22,40 @@
...
+  bool debug = false;
```
Comment (dzelemba) [inline src/wsiToDcm.h]: Add comment.

File: src/wsiToDcm.h [hunk 12]
```diff
@@ -22,51 +22,40 @@
 
 namespace wsiToDicomConverter {
 
+// Structure for wsi2dcm settings
+struct WsiRequest {
+  std::string inputFile = "";  // wsi input
```
[the remark this answers is not in the document]
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/wsiToDcm.h [hunk 13]
```diff
@@ -22,51 +22,40 @@
...
+  std::string outputFileMask = "./";  // path to save generated files
+  int64_t frameSizeX = 500;
+  int64_t frameSizeY = 500;
+  DCM_Compression compression = JPEG;  // compression - jpeg, jpeg2000, raw
+  int32_t quality = 80;
```
[the remark this answers is not in the document]
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/wsiToDcm.h (hunk 6 above)
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/wsiToDcm.h (hunk 7 above)
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/wsiToDcm.h (hunk 8 above)
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/wsiToDcm.h (hunk 10 above)
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/wsiToDcm.h (hunk 11 above)
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/enums.h [hunk 14]
```diff
@@ -15,6 +15,25 @@
...
+#include <algorithm>
+#include <string>
+
 typedef enum { UNKNOWN = -1, JPEG2000 = 0, JPEG = 1, RAW = 2 } DCM_Compression;
 
+inline DCM_Compression dcmCompressionFromString(std::string compressionStr) {
+  DCM_Compression compression = UNKNOWN;
+  std::transform(compressionStr.begin(), compressionStr.end(),
+                 compressionStr.begin(), ::tolower);
+  if (compressionStr.find("jpeg") == 0) {
```
Comment (dzelemba) [inline src/enums.h]: Won't this trigger for "jpeg2000"? Why use find() instead of equals()? Also can we add a unit test for this method?

Comment (pavertomato) [reply src/enums.h]: > Why use find() instead of equals()
do you meant compare()? to cover additional chars what may be entered by user
> Won't this trigger for "jpeg2000"?
yes, but would be overridden, since it called one time per file i didn't put jpeg200 if into jpeg if

File: src/wsiToDcm.h (hunk 9 above)
Comment (pavertomato) [reply src/wsiToDcm.h]: done

File: src/enums.h (hunk 14 above)
Comment (dzelemba) [reply src/enums.h]: What is someone puts in a variant of JPEG we don't support? E.g. jpegls or some other obscure one. They might think they're getting lossless compression, but we'll just be using lossy jpeg.

I think we should be more strict here.

Comment (pavertomato) [reply src/enums.h]: done
